src/v/archival/upload_housekeeping_service.h

```diff
@@ -0,0 +1,221 @@
+/// the idle timer will never be triggered and its callback will never be
+/// called.
+///
+/// The 'epoch' timer runs less frequently. Notifications from the cloud
+/// storage do not reset it. So it will be triggered eventually if the cloud
+/// storage is always busy. The callback of the epoch timer initiates transition
+/// to the active state. It's purpose is to make sure that the housekeeping job
+/// is perfomed even when the cloud storage is always busy. This will drain the
+/// backlog (the job queue) so it's guaranteed that every job is executed at
+/// least once per epoch.
+///
+/// The noticifactions from the cloud storage API reset the idle timer so if the
+/// cloud storage is busy it will never be triggered. If any of the jobs uses
+/// the cloud storage API the notifications from it will be ignored.
+///
+/// The service might be stopped by calling 'stop' method. In this case method
+/// 'interrupt' of the job will be called if there is a job which is executing
+/// at the moment. The 'ntp_archiver_service' has to deregister and stop its
+/// job before deletion.
+///
+/// The following configuration parameters are used:
+/// - cloud_storage_idle_timeout_ms: the service goes into active state when
+///   remote api is not active for cloud_storage_idle_timeout_ms milliseconds.
+/// - cloud_storage_housekeeping_interval_ms: epoch timeout. The service goes
+///   active state forcibly, even if remote api is not idle.
+/// - cloud_storage_segment_upload_timeout x 3: if service is stopping being
+///   active for longer than 3x segment upload timeout the state is stopped
+///   forcibly and the error is generated.
+class upload_housekeeping_service {
+public:
+    explicit upload_housekeeping_service(
+      ss::sharded<cloud_storage::remote>& api);
+
+    upload_housekeeping_service(const upload_housekeeping_service&) = delete;
+    upload_housekeeping_service(upload_housekeeping_service&&) = delete;
+    upload_housekeeping_service& operator=(const upload_housekeeping_service&)
+      = delete;
+    upload_housekeeping_service& operator=(upload_housekeeping_service&&)
+      = delete;
+
+    ~upload_housekeeping_service();
+
+    /// Register housekeeping jobs
+    void
+    register_jobs(std::vector<std::reference_wrapper<housekeeping_job>> jobs);
+
+    void
+    deregister_jobs(std::vector<std::reference_wrapper<housekeeping_job>> jobs);
+
+    /// Start/stop the service
+    ss::future<> start();
+    ss::future<> stop();
+
+    /// Return workflow object
+    housekeeping_workflow& workflow();
+
+private:
+    ss::future<> bg_idle_loop();
+    void rearm_idle_timer();
+    void idle_timer_callback();
+    void epoch_timer_callback();
+
+private:
+    ss::gate _gate;
+    ss::abort_source _as;
+    /// Cloud storage API
+    cloud_storage::remote& _remote;
+    /// Idle timer, used to detect idle state
+    ss::timer<ss::lowres_clock> _idle_timer;
+    ss::timer<ss::lowres_clock> _epoch_timer;
```

Comment:
nit: maybe we can call this `_force_housekeep_timer` or somesuch? "epoch" is a bit too general and doesn't quite capture the context without already understanding how it works IMO